drivers/adc/adc_max11102_17.c

```diff
@@ -0,0 +1,460 @@
+{
+	int result;
+	struct max11102_17_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	result = max11102_17_adc_start_read(dev, sequence, false);
+
+	while (result == 0 && k_sem_take(&data->ctx.sync, K_NO_WAIT) != 0) {
+		result = max11102_17_adc_perform_read(dev);
+	}
+
+	adc_context_release(&data->ctx, result);
+	return result;
+}
+#endif
+
+#if CONFIG_ADC_ASYNC
+static void max11102_17_acquisition_thread(struct device *dev)
+{
+	while (true) {
+		max11102_17_adc_perform_read(dev);
+	}
+}
+#endif
+
+static int max11102_17_init(const struct device *dev)
+{
+	int result;
+	const struct max11102_17_config *config = dev->config;
+	struct max11102_17_data *data = dev->data;
+	int16_t sample;
+
+	adc_context_init(&data->ctx);
+
+	k_sem_init(&data->acquire_signal, 0, 1);
+
+	if (!spi_is_ready_dt(&config->bus)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	switch (config->channel_count) {
+	case 1:
+		if (config->gpio_chsel.port != NULL) {
+			LOG_ERR("GPIO for chsel set with only one channel");
+			return -EINVAL;
+		}
+		break;
+	case 2:
+		if (config->gpio_chsel.port == NULL) {
+			LOG_ERR("no GPIO for chsel set with two channels");
+			return -EINVAL;
+		}
+
+		result = gpio_pin_configure_dt(&config->gpio_chsel, GPIO_OUTPUT_INACTIVE);
+		if (result != 0) {
+			LOG_ERR("failed to initialize GPIO for chsel");
+			return result;
+		}
+		break;
+	default:
+		LOG_ERR("invalid number of channels (%i)", config->channel_count);
+		return -EINVAL;
+	}
+
+	data->current_channel_id = 0;
+
+#if CONFIG_ADC_ASYNC
+	const k_tid_t tid = k_thread_create(
+		&data->thread, config->stack, CONFIG_ADC_MAX11102_17_ACQUISITION_THREAD_STACK_SIZE,
+		(k_thread_entry_t)max11102_17_acquisition_thread, (void *)dev, NULL, NULL,
+		CONFIG_ADC_MAX11102_17_ASYNC_THREAD_INIT_PRIO, 0, K_NO_WAIT);
+	k_thread_name_set(tid, "adc_max11102_17");
+#endif
+
+	/* power up time is one conversion cycle */
+	result = max11102_17_read_sample(dev, &sample);
+	if (result != 0) {
+		LOG_ERR("unable to read dummy sample for power up timing");
+		return result;
+	}
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return result;
+}
+
+static const struct adc_driver_api api = {
+	.channel_setup = max11102_17_channel_setup,
+	.read = max11102_17_read,
+	.ref_internal = 0,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = max11102_17_adc_read_async,
+#endif
+};
+
+BUILD_ASSERT(CONFIG_ADC_INIT_PRIORITY > CONFIG_SPI_INIT_PRIORITY,
+	     "CONFIG_ADC_INIT_PRIORITY must be higher than CONFIG_SPI_INIT_PRIORITY");
+
+#define ADC_MAX11102_17_INST_DEFINE(index, name, res, channels)                                    \
+	IF_ENABLED(CONFIG_ADC_ASYNC,                                                               \
+		   (static K_KERNEL_STACK_DEFINE(                                                  \
+			    thread_stack_##name##_##index,                                         \
+			    CONFIG_ADC_MAX11102_17_ACQUISITION_THREAD_STACK_SIZE);))               \
```

Comment:
You can use `K_KERNEL_STACK_MEMBER` to define this within `struct max11102_17_data`

Reply:
done